source/change-streams/change-streams.rst

```diff
@@ -414,25 +462,96 @@ Once a ``ChangeStream`` has encountered a resumable error, it MUST attempt to re
+  - In this case, the ``ChangeStream`` will return notifications starting with the oplog entry immediately *after* the provided token.
 
 If the server supports sessions, the resume attempt MUST use the same session as the previous attempt's command.
 
 A driver SHOULD attempt to kill the cursor on the server on which the cursor is opened during the resume process, and MUST NOT attempt to kill the cursor on any other server.
 
 
+Exposing All Resume Tokens
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:since: 4.2
+
+Users can retrieve the ``documentResumeToken`` by inspecting the _id on each ``ChangeDocument``. But since MongoDB 4.2, aggregate and getMore responses also include a ``postBatchResumeToken``. Drivers use one or the other when automatically resuming, as described in `Resume Process`_.
+
+Drivers MUST expose a mechanism to retrieve the same resume token that would be used to automatically resume. It MUST be possible to use this mechanism after iterating every document *and* after every response to a getMore (including empty responses). Drivers have two options to implement this.
+
+Option 1: ChangeStream::getResumeToken()
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code:: typescript
+
+  interface ChangeStream extends Iterable<Document> {
+    /**
+     * Returns a resume token that should be used to resume after the most
+     * recently returned change.
+     */
+    public getResumeToken() Optional<Document>;
+  }
+
+This method returns the cached ``documentResumeToken`` or ``postBatchResumeToken`` following rules from `Resume Process`_:
+
+- If the ``ChangeStream`` has no cached ``documentResumeToken`` or ``postBatchResumeToken`` this returns an unset optional.
+- If the ``ChangeStream`` has a cached ``postBatchResumeToken`` and has returned all documents in the most recent batch (or the most recent batch was empty) this returns the ``postBatchResumeToken``.
+- Otherwise, this returns the cached ``documentResumeToken``.
+
+This MUST be implemented in synchronous drivers. This MAY be implemented in asynchronous drivers.
+
+Option 2: Event Emitted for Resume Token
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Allow users to set a callback to listen for new resume tokens. The exact interface is up to the driver, but it MUST meet these criteria:
+
+- The callback is set in the same manner as a callback used for receiving change documents.
+- The callback accepts a resume token as an argument.
+- The callback (or event) MAY include an optional ChangeDocument, which is unset when called with resume tokens sourced from ``postBatchResumeToken``.
+
+A possible interface for this callback MAY look like:
+
+.. code:: typescript
+
+  interface ChangeStream extends Iterable<Document> {
+    /**
+     * Returns a resume token that should be used to resume after the most
+     * recently returned change.
+     */
+    public onResumeTokenChanged(ResumeTokenCallback:(Document resumeToken) => void);
```

Comment:
I thought we decided for this option to return change documents and resume tokens together?

Reply:
This is an example and not the required interface. I chose to leave it less strictly specified because I think it might be preferable in some drivers to have a separate event just for resume token updates. For example, in NodeJS, I've checked with @mbroadst that it would not be necessary to have both the document and resume token as part of the callback arguments.